File: Go/054. Spiral Matrix.go
package main

func spiralOrder(matrix [][]int) []int {
	res := make([]int, 0)
	if len(matrix) == 0 || len(matrix[0]) == 0 {
		return res
	}
	row, col := len(matrix), len(matrix[0])
	seen := make([][]bool, row)
	for i := range seen {
		seen[i] = make([]bool, col)
	}
	addRow, addCol := []int{0, 1, 0, -1}, []int{1, 0, -1, 0}
	i, j, add := 0, 0, 0
	for k := 0; k < row * col; k++ {
		res = append(res, matrix[i][j])
		seen[i][j] = true
		nextI, nextJ := i + addRow[add], j + addCol[add]
		condition1 := 0 <= nextI && nextI < row
		condition2 := 0 <= nextJ && nextJ < col
		if condition1 && condition2 && !seen[nextI][nextJ]{
			i, j = nextI, nextJ
		} else {
			add = (add + 1) % 4
			i, j = i + addRow[add], j + addCol[add]
		}
	}
	return res
}

File: C++/054. Spiral Matrix.cpp
#include <iostream>
#include <vector>

class Solution {
public:
    std::vector<int> spiralOrder(std::vector<std::vector<int>>& matrix) {
        std::vector<int> res;
        if (matrix.size() == 0 || matrix[0].size() == 0)
            return res;
        int row = int(matrix.size()), col = int(matrix[0].size());
        std::vector<std::vector<bool>> seen(row, std::vector<bool>(col, false));
        std::vector<int> add_row{0, 1, 0, -1}, add_col{1, 0, -1, 0};
        int i = 0, j = 0, add = 0;
        for (int k = 0; k < row * col; k++) {
            res.emplace_back(matrix[i][j]);
            seen[i][j] = true;
            int next_i = i + add_row[add], next_j = j + add_col[add];
            bool condition1 = 0 <= next_i && next_i < row;
            bool condition2 = 0 <= next_j && next_j < col;
            if (condition1 && condition2 && !seen[next_i][next_j]) {
                i = next_i;
                j = next_j;
            } else {
                add = (add + 1) % 4;
                i = i + add_row[add];
                j = j + add_col[add];
            }
        }
        return res;
    }
};